apps/emqx/src/emqx_limiter/src/emqx_limiter_container.erl

```diff
@@ -0,0 +1,141 @@
+
+-define(RETRY_KEY(Type), {retry, Type}).
+-type retry_key() :: ?RETRY_KEY(limiter_type()).
+
+%%--------------------------------------------------------------------
+%%  API
+%%--------------------------------------------------------------------
+-spec new() -> container().
+new() ->
+    new([]).
+
+%% @doc generate default data according to the type of limiter
+-spec new(list(limiter_type())) -> container().
+new(Types) ->
+    new(Types, #{}).
+
+%% @doc generate a container
+%% according to the type of limiter and the bucket name configuration of the limiter
+%% @end
+-spec new(list(limiter_type()), #{limiter_type() => emqx_limiter_schema:bucket_name()}) -> container().
+new(Types, Buckets) ->
+    Init = fun(Type, Acc) ->
+                   Client = emqx_limiter_server:connect(Type, Buckets),
+                   add(Type, Client, Acc)
+           end,
+    lists:foldl(Init, #{}, Types).
+
+%% @doc add the specified type of client to the container
+-spec add_by_name(limiter_type(),
+                  bucket_name() | #{limiter_type() => bucket_name()},
+                  container()) -> container().
+add_by_name(Type, Buckets, Container) ->
+    Client = emqx_limiter_server:connect(Type, Buckets),
+    add(Type, Client, Container).
+
+-spec add(limiter_type(), client(), container()) -> container().
+add(Type, Client, Container) ->
+    Container#{ Type => Client
+              , ?RETRY_KEY(Type) => undefined
+              }.
+
+%% @doc check the specified client
+-spec check(pos_integer(), limiter_type(), container()) -> check_result().
+check(Need, Type, C) ->
+    check_list([{Need, Type}], C).
+
+%% @doc check multiple clients
+-spec check_list(list({pos_integer(), limiter_type()}), container()) -> check_result().
+check_list([{Need, Type} | T], C) ->
+    Client = maps:get(Type, C),
+    case emqx_limiter_client:check(Need, Client) of
+        {ok, Client2} ->
+            check_list(T, C#{Type := Client2});
+        {_, Pause, Ctx, Client2} ->
+            Fun = fun({FN, FT}, Acc) ->
+                          Future = emqx_limiter_client:make_future(FN),
+                          Acc#{?RETRY_KEY(FT) := Future}
+                  end,
+            C2 = lists:foldl(Fun,
+                             C#{Type := Client2,
+                                ?RETRY_KEY(Type) := Ctx},
+                             T),
+            {pause, Pause, C2}
+    end;
+
+check_list([], C) ->
+    {ok, C}.
+
+%% @doc retry the specified client
+-spec retry(limiter_type(), container()) -> check_result().
+retry(Type, C) ->
+    retry_list([Type], C).
+
+%% @doc retry multiple clients
+-spec retry_list(list(limiter_type()), container()) -> check_result().
+retry_list([Type | T], C) ->
+    Key = ?RETRY_KEY(Type),
+    case C of
+        #{Type := Client,
+          Key := Retry} when Retry =/= undefined ->
+            case emqx_limiter_client:check(Retry, Client) of
+                {ok, Client2} ->
```

Comment:
add comments about why we do not delete the `Key`, why set `Key` (the RetryKey) to `undefined.` and describe why the `undefined` is used.